spring-kafka/src/main/java/org/springframework/kafka/support/serializer/ErrorHandlingDeserializer2.java

```diff
@@ -185,9 +199,27 @@ public T deserialize(String topic, Headers headers, byte[] data) {
 		}
 		catch (Exception e) {
 			deserializationException(headers, data, e);
-			return this.failedDeserializationFunction != null
-					? this.failedDeserializationFunction.apply(data, headers)
-					: null;
+			return recoverFromSupplier(topic, headers, data, e);
+		}
+	}
+
+	private T recoverFromSupplier(String topic, Headers headers, byte[] data, Exception exception) {
+		if (this.failedDeserializationFunction != null) {
+			FailedDeserializationInfo t = new FailedDeserializationInfo(topic, headers, data, this.isForKey, exception);
+			return this.failedDeserializationFunction.apply(t);
+		}
+		else {
+			return null;
+		}
+	}
+
+	private T recoverFromSupplier(String topic, byte[] data, Exception exception) {
```

Comment:
Why do we need two of these methods?
Why single one with broader context can't cover us?